### DISCUSSION TITLE
WASM examples

### DISCUSSION THREAD

**Discussion by elalish** [elalish/manifold#158] [Ideas]
@makc @pca006132 I really want to serve some web examples on our Github pages so people can interact with our library live. I think the example with three.js is a perfect place to start. All we need is to make building that WASM library part of our `deploy.yml` script and add relevant files to `deploy.sh` to get them served. 

~~One question: do we still need our `build_wasm` CI? It includes Assimp and our tests and everything - I wonder if this is really necessary any more? It seems a little odd to have two different WASM builds, but do we think it's important to independently test it? Curious what your thoughts are here.~~

**Comment by pca006132** [comment]
The idea is just to test the build script on two platforms, nix and ubuntu, to avoid any changes from breaking these two builds. I guess I am the only user for nix here, and most of the users will follow the build instructions for ubuntu. I'm fine removing the nix one.

**Reply by elalish** [reply]
Sorry, I wasn't referring to nix; that's fine. I was just thinking it was odd that we build two different WASM bundles now - one for testing and a different one for binding.

**Reply by pca006132** [reply]
Indeed, but I think it is fine: we don't want to include the testing stuff in the binding wasm, as we want a smaller binary.

**Comment by makc** [comment]
@elalish if you comment assimp and gtest out, wasm builds just fine. but tests use meshio, so those fail 🤷

**Comment by makc** [comment]
@pca006132 if you remove [these lines](https://github.com/elalish/manifold/blob/master/.github/workflows/manifold.yml#L99-L102), shouldnt you be able to successfully run with the code I had [here](https://github.com/elalish/manifold/pull/151/commits/7022c0f4c30cb07e129985ab825892b6cad3dade)?

**Reply by pca006132** [reply]
The tests and bindings are built separately, so yes we should be able to run your binding code. However, we don't need to remove the test code in the CI, as they are not included in the binding wasm file, they just make the CI take a bit longer which is acceptable.

**Reply by makc** [reply]
but it literally fails the build if you do that and not remove the test - which is why I reverted the change

**Reply by makc** [reply]
(by "fails" I mean it shows up red in github. the actual build works if you do it by hand)

**Reply by pca006132** [reply]
I think it probably failed the nix build, as we execute the tests on nix also. I think keeping the test is good as we want to make sure wasm is not doing something unexpected.

**Reply by makc** [reply]
nix one was a separate issue that you have fixed. I am not an expert, but it kind of sounds strange that you say the test would still work when I disable test folders in CMakeLists 🤣

**Reply by pca006132** [reply]
Oh ok, I did not actually run it (I'm lying on my bed now) but I thought it should be fine. I will have a look at it tmr.

**Reply by makc** [reply]
Wait, I'm still processing. Does @elalish mean that if you remove the entire [build_wasm](https://github.com/elalish/manifold/blob/master/.github/workflows/manifold.yml#L71) section, js demo will still be built elsewhere?

**Reply by makc** [reply]
...is build_nix (js) it?

**Reply by elalish** [reply]
Honestly, I'm still trying to make sense of this myself. Let's move this discussion to the [PR](https://github.com/elalish/manifold/pull/159) so we have some context. Also, @makc did you notice the big directory reorganization we just did? Your wasm stuff is now under `bindings`.

**Comment by elalish** [comment]
It works! Awesome team effort, all! https://elalish.github.io/manifold/bindings/wasm/examples/index.html

**Reply by fire** [reply]
On my firefox client the page does not load.

**Reply by pca006132** [reply]
Works fine on my computer (firefox 102.0.1). Can you open the network panel and see if it is caused by some scripts not being load or js/wasm error?

**Reply by fire** [reply]
Somehow I'm getting a https://unpkg.com/three url and it's not loading.

**Reply by makc** [reply]
it does indeed load two files from there:
* https://unpkg.com/three@0.142.0/build/three.js
* https://unpkg.com/three@0.142.0/examples/js/utils/BufferGeometryUtils.js

Idk why would it not load for @fire but for me it loads fine in firefox. maybe it is network issue.

**Comment by stewartoallen** [comment]
excited to have been pointed to this project! manifoldcad.org works but the [simple demo being served here](https://manifoldcad.org/three.html) does not.

![Screen Shot 2022-10-21 at 1 23 05 PM](https://user-images.githubusercontent.com/554275/197253823-26ff5ade-aad4-4ea7-a7a2-831b0fc46413.png)

**Reply by elalish** [reply]
Oh, good call, I bet I know exactly how I broke that yesterday. Will fix.

**Reply by elalish** [reply]
Fixed.
